File: sdk-node/examples/invoices/refund-payment.ts
/**
 * Refund all or part of a recorded payment on a paid invoice. The
 * `idempotencyKey` makes the request safe to replay — refunding twice with the
 * same key returns the existing refund instead of issuing a second one.
 *
 * Run:
 *   npx tsx examples/invoices/refund-payment.ts
 */

import { ThreeCommon } from '@3common/sdk'

const client = new ThreeCommon({
  apiKey: '3co_your_api_key_here',
})

const refunded = await client.invoices.refundPayment(
  'inv_replace_with_real_id',
  'pay_replace_with_real_id',
  {
    amount: 25_000, // $250.00 in cents; capped at the payment's refundable balance
    reason: 'requested_by_customer',
    // Derive the idempotency key from a stable business event id (e.g. the
    // refund-request id in your own system) — never the wall clock. A fresh
    // value on each run is a new key, so a retry after a crash refunds twice.
    idempotencyKey: 'rfnd-8842',
  },
)

console.log(`invoice ${refunded.id ?? '?'} now ${refunded.status ?? '?'}`)
console.log(`  paid: ${String(refunded.amountPaid ?? 0)}, due: ${String(refunded.amountDue ?? 0)}`)

File: sdk-python/examples/invoices/refund_payment.py
"""Refund all or part of a recorded payment on a paid invoice.

The ``idempotency_key`` makes the request safe to replay — refunding twice with
the same key returns the existing refund instead of issuing a second one.

Run with: python examples/invoices/refund_payment.py
"""

from __future__ import annotations

from threecommon import ThreeCommon
from threecommon.invoices import RefundBody


def main() -> None:
    with ThreeCommon(api_key="3co_your_api_key_here") as client:
        # Derive the idempotency key from a stable business event id (e.g. the
        # refund-request id from your own system) — never the wall clock. A fresh
        # timestamp on each run is a *new* key, so a retry after a crash would
        # issue a second refund instead of replaying the first.
        refunded = client.invoices.refund_payment(
            "inv_replace_with_real_id",
            "pay_replace_with_real_id",
            RefundBody(
                amount=25_000,  # $250.00 in cents; capped at the refundable balance
                reason="requested_by_customer",
                idempotency_key="rfnd-replace_with_business_event_id",
            ),
        )
        print(f"invoice {refunded.id} now {refunded.status}")
        print(f"  paid: {refunded.amount_paid}, due: {refunded.amount_due}")


if __name__ == "__main__":
    main()
